smwe_rom/src/snes_utils/rom_slice.rs

```diff
@@ -0,0 +1,76 @@
+use std::{fmt, fmt::Formatter};
+
+use crate::snes_utils::addr::{Addr, AddrPc, AddrSnes};
+
+#[derive(Copy, Clone, Debug)]
+pub struct RomSlice<A: Addr> {
+    pub begin: A,
+    pub size:  usize,
+}
+
+pub type PcSlice = RomSlice<AddrPc>;
+pub type SnesSlice = RomSlice<AddrSnes>;
+
+impl PcSlice {
+    pub const fn new(begin: AddrPc, size: usize) -> Self {
+        Self { begin, size }
+    }
+}
+
+impl SnesSlice {
+    pub const fn new(begin: AddrSnes, size: usize) -> Self {
+        Self { begin, size }
+    }
+}
+
+impl<A: Addr> RomSlice<A> {
+    pub fn end(&self) -> A {
+        self.begin + self.size
+    }
+
+    pub fn shift_forward(self, offset: usize) -> Self {
+        Self { begin: self.begin + offset, ..self }
+    }
+
+    pub fn shift_backward(self, offset: usize) -> Self {
+        Self { begin: self.begin - offset, ..self }
+    }
+
+    pub fn skip_forward(self, times_size: usize) -> Self {
+        Self { begin: self.begin + (self.size * times_size), ..self }
+    }
+
+    pub fn skip_backward(self, times_size: usize) -> Self {
+        Self { begin: self.begin - (self.size * times_size), ..self }
+    }
+
+    pub fn move_to(self, new_address: A) -> Self {
+        Self { begin: new_address, ..self }
+    }
+
+    pub fn expand(self, diff: usize) -> Self {
+        Self { size: self.size + diff, ..self }
+    }
+
+    pub fn shrink(self, diff: usize) -> Self {
+        Self { size: self.size - diff, ..self }
+    }
+
+    pub fn resize(self, new_size: usize) -> Self {
+        Self { size: new_size, ..self }
+    }
+
+    pub fn infinite(self) -> Self {
```

Comment:
If you want to represent infinite slices, either make it a separate type like Rust stdlib does, use enums for state, or at least use `usize::MAX` if you want to reserve a special value. Using zero here doesn't make a lot of sense as zero-size slices are entirely valid and this way you're introducing very surprising behaviour. If you want to reserve a special state, make sure that the functions do some basic error checking so that e.g. expanding an infinite slice does nothing or returns None.